add discoverer component overview document (#1574)
vdaas/vald: docs/overview/component/discoverer.md

# Vald Discoverer

Vald Discoverer runs individually and helps automate operations for the Vald cluster.

It is also the only component that communicates with [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).

## Responsibility

> Vald Discoverer is responsible for sharing resource usage about each Node and Pod to the other components by communicating [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).

Reviewer: ```suggestion
Vald Discoverer is responsible to retrieve the resource usage of each Node and Pod from [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) and share it with other components in Vald cluster.
```

Author: ```suggestion
Vald Discoverer is responsible for retrieving each Node and Pod resource usage from [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) and sharing it with other components in the Vald cluster.
```

## Feature

### Getting Node and Pod metrics

> Vald Discoverer requires [kube-apiserver](<[https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/](https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/)>) for getting each Node and Pod metrics in the Vald cluster.

Reviewer: ```suggestion
Vald Discoverer requires [kube-apiserver](<[https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/](https://kubernetes.io/ja/docs/reference/command-line-tools-reference/kube-apiserver/)>) to get Node and Pod metrics in the Vald cluster.
```

> It uses `Reconcile function`, one of the internal implementations, which runs on changing metrics for syncing metrics.

Reviewer: ```suggestion
It synchronizes the metrics when it changes by implementing the `Reconcile function`.
```

  | Pods         | Pod information in the node     |

- Pod

  | Key name  | Description                     |
  | --------- | ------------------------------- |
  | AppName   | AppName of pod                  |
  | Name      | pod name                        |
  | Namespace | namespace where pod runs        |
  | IP        | IP address                      |
  | CPU       | CPU limit and CPU request       |
  | Memory    | Memory limit and Memory request |

When syncing success, Vald Discoverer chooses the necessary metrics from the result and stores them into their four kinds of `Map` on their local memory:

| Map name        | Description                          |
| --------------- | ------------------------------------ |
| podsByNode      | pod metrics list group by node       |
| podsByNamespace | pod metrics list group by namespace  |
| podsByName      | pod metrics list group by pod name   |
| nodebyName      | node metrics list group by node name |

<!-- TODO:image -->

### Sharing Node and Pod metrics

Vald Discoverer provides the internal client for another component.

The component, which requires the metrics information, uses when it needs.

> In the Vald cluster, two components, Vald LB Gateway and Vald Manager Index, require for running their functions.

Reviewer: ```suggestion
Vald LB Gateway and Vald Manager Index in the Vald cluster require it to support their features.
```

Author: `support` is a weak word in this context, because each component must use it.
IMO, there would be better.
How about?

suggestion 1
```suggestion
Vald LB Gateway and Vald Manager Index in the Vald cluster require it to achieve its responsibility.
```

suggestion 2
```suggestion
Vald LB Gateway and Vald Manager Index in the Vald cluster require it to run its function.
```

Reviewer: I think 1 is better. Thank you for your response :)
